pkg/quarterdeck/db/tokens.go

```diff
@@ -57,29 +52,76 @@ func (t *VerificationToken) Sign() (_ string, secret []byte, err error) {
 		return "", nil, err
 	}
 
-	// Compute hash with a random 64 byte key
-	key := make([]byte, keyLength)
-	if _, err = rand.Read(key); err != nil {
-		return "", nil, err
+	return t.signData(data)
+}
+
+// Verify checks that a token was signed with the secret and is not expired.
+func (t *VerificationToken) Verify(signature string, secret []byte) (err error) {
+	if t.Email == "" {
+		return ErrTokenMissingEmail
 	}
 
-	mac := hmac.New(sha256.New, key)
-	if _, err = mac.Write(data); err != nil {
-		return "", nil, err
+	if t.IsExpired() {
+		return ErrTokenExpired
 	}
 
-	// Include the nonce with the key so that the token can be reconstructed later
-	secret = make([]byte, nonceLength+keyLength)
-	copy(secret[0:nonceLength], t.Nonce)
-	copy(secret[nonceLength:], key)
+	if len(secret) != nonceLength+keyLength {
+		return ErrInvalidSecret
+	}
 
-	return base64.RawURLEncoding.EncodeToString(mac.Sum(nil)), secret, nil
+	// Serialize the struct with the nonce from the secret
+	t.Nonce = secret[0:nonceLength]
+	var data []byte
+	if data, err = msgpack.Marshal(t); err != nil {
+		return err
+	}
+
+	return t.verifyData(data, signature, secret)
+}
+
+// NewResetToken creates a token struct from a user ID that expires in 15 minutes.
+func NewResetToken(id ulid.ULID) (token *ResetToken, err error) {
+	if ulids.IsZero(id) {
+		return nil, ErrMissingUserID
+	}
+
+	token = &ResetToken{
+		UserID: id,
+	}
+
+	if token.SigningInfo, err = NewSigningInfo(time.Minute * 15); err != nil {
+		return nil, err
+	}
+
+	if _, err = rand.Read(token.Nonce); err != nil {
+		return nil, fmt.Errorf("could not generate token: %w", err)
+	}
+	return token, nil
+}
+
+// ResetToken packages a user ID with random data and an expiration time so that it can
+// be serialized and hashed into a token which can be sent to users.
+type ResetToken struct {
```

Comment:
I was thinking of reusing the email verification fields in the database to store the secret, since I think we will not be sending reset password emails to unverified emails right?